### PR TITLE
Fix FAT bugs and increase file count limit

### PR DESCRIPTION
**Pull request by henrygab**
Five categories
1. FAT requires directory entries to have (at minimum) the creation date
2. Allow more than 15 files in root directory
3. Add some STATIC_ASSERT() where possible, comment when not possible
4. Add comments on unwritten presumptions of the code
5. Use ARRAY_SIZE2() macro

## FAT Directory Entry bug
Attempting to list files from a CMD window under Windows 10 failed.  Tab filename completion failed. The root cause was the directory entries created by ghostfat.c failed to fill in the directory creation date/time … a value of zero is invalid.  See fatgen103.pdf, which states:

> Many FAT file systems do not support Date/Time other than DIR_WrtTime and DIR_WrtDate. For this reason, DIR_CrtTimeMil, DIR_CrtTime, DIR_CrtDate, and DIR_LstAccDate are actually optional fields. DIR_WrtTime and DIR_WrtDate must be supported, however.

I chose to use the __DATE__ and __TIME__ preprocessor macros, so that the filesystem files would always appear to be created at the date/time that the bootloader was built.  The preprocessor does all the work, so only literal values are emitted into the code … no runtime string manipulation.

## Increased file limits
Initially, I simply added STATIC_ASSERT() that all the files would fit in a single sector.  This limits the root file system to only 15 files.
Because ghostfat.c already reserved four sectors for the root directory, it was fairly easy to update the code to support >15 files (but still limited to 62).   Exceeding the limit will cause a compile-time assertion.
For ease of testing, I added a [```#define```](https://github.com/adafruit/Adafruit_nRF52_Bootloader/commit/53a99195bfc3aa549268f95bb0ecef401cf65c77#diff-f5fb77133f169c7ff0c1985633e0e24dR12) that adds an extra 20 files to the root directory.

## STATIC_ASSERT() additions
Static assertions catch code errors before they even get loaded onto the chip, and document presumptions in the code.  Enough said.

## Comments on unwritten presumptions of the code
While not complete, it's hoped that some of these comments will help avoid future UF2 implementations from breaking some of the "unwritten rules" that exist in ghostfat.c.

## Use ARRAY_SIZE2() macro
The C macro ```#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(array[0]))``` is not type safe.  If the argument is ever changed to a pointer, the macro will **_silently_** return ```sizeof(uintptr_t) / sizeof(arr[0])```, which is often zero and rarely correct.

I could not discover a type-safe solution for C99 code.  However, it might be possible to someday compile this code as C++, which would now automatically update to type-safe, constexpr versions..

### CODE REVIEW

File: src/usb/uf2/ghostfat.c [hunk 1]
```diff
@@ -200,26 +272,49 @@ void read_block(uint32_t block_no, uint8_t *data) {
...
+            // DIR_WrtTime and DIR_WrtDate must be supported
+            d->updateTime       = __DOSTIME__;
+            d->updateDate       = __DOSDATE__;
+            d->startCluster     = startCluster & 0xFF;
+            // WARNING -- code presumes only one NULL .content for .UF2 file
+            //            and requires it be the last element of the array
+            d->size = inf->content ? strlen(inf->content) : UF2_SIZE;
+        }
+
+    } else {
```
Comment (hathach) [inline src/usb/uf2/ghostfat.c]: `// else Generate the UF2 file data on-the-fly`
I really like your comment that further explain the code in previous PR , why would you want to undo it in PR. I can add it back, no problems at all. Just wonder your reason

File: src/usb/uf2/macros.h [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    // the third line.
+    // Moreover, check1 and check2 are reduced to the value zero,
+    // while the third line is the old type-unsafe C-style macro,
+    // now made entirely type-safe.
+    // 
+    // Additional benefits:
+    // The result is itself constexpr
+    // 
+    //
+    #define ARRAY_SIZE2(arr) ( \
```
Comment (hathach) [inline src/usb/uf2/macros.h]: This is lots of thought, thank you very much for bringing this up :+1:

File: src/usb/uf2/ghostfat.c [hunk 3]
```diff
@@ -59,12 +63,14 @@ struct TextFile {
...
 #define STR(x) STR0(x)
+
 const char infoUf2File[] = //
     "UF2 Bootloader " UF2_VERSION "\r\n"
     "Model: " PRODUCT_NAME "\r\n"
     "Board-ID: " BOARD_ID "\r\n"
     "Bootloader: " BOOTLOADER_ID "\r\n"
-    "Date: " __DATE__ "\r\n";
+    "Date: " __DATE__ "\r\n"
+    "Time: " __TIME__ "\r\n";
```
[the remark this answers is not in the document]
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: Ok, removed.

File: src/usb/uf2/ghostfat.c [hunk 4]
```diff
@@ -76,16 +82,72 @@ const char indexFile[] = //
     "</body>"
     "</html>\n";
 
+#ifdef CREATE_MANY_FILES
```
[the remark this answers is not in the document]
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: I disagree with this request. I find built-in test code that is controlled by a single #define to be important, as it simplifies validating that later changes do not break functionality.
However, this is your project, so I've complied with the request.

File: src/usb/uf2/ghostfat.c [hunk 5]
```diff
@@ -9,6 +11,9 @@
 #include "bootloader_settings.h"
 #include "bootloader.h"
 
+// Uncomment the next line to create an additional 20 files in root.
+// #define CREATE_MANY_FILES
```
[the remark this answers is not in the document]
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: I disagree with this request. I find built-in test code that is controlled by a single #define to be important, as it simplifies validating that later changes do not break functionality.
However, this is your project, so I've complied with the request.

File: src/usb/uf2/ghostfat.c [hunk 6]
```diff
@@ -190,8 +259,11 @@ void read_block(uint32_t block_no, uint8_t *data) {
         if (sectionIdx >= SECTORS_PER_FAT)
             sectionIdx -= SECTORS_PER_FAT; // second FAT is same as the first...
         if (sectionIdx == 0) {
-            data[0] = 0xf8; // first FAT entry must match BPB MediaDescriptor
-            for (int i = 1; i < NUM_INFO * 2 + 4; ++i) {
+            data[0] = 0xf0;
```
[the remark this answers is not in the document]
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: I keep each patch branch to the smallest possible subset of changes.  This patch branch has not merged those prior changes.  I will work to merge them now.

File: src/usb/uf2/ghostfat.c (hunk 1 above)
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: I try to keep each patch branch focused on stand-alone changes.  This patch branch has not merged those prior changes.  I will work to merge them now.

File: src/usb/uf2/macros.h (hunk 2 above)
Comment (henrygab) [reply src/usb/uf2/macros.h]: my pleasure.

File: src/usb/uf2/ghostfat.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: ok, manually added those changes.   not sure why they got lost, I thought GIT would be smart enough to merge them properly.

File: src/usb/uf2/ghostfat.c (hunk 1 above)
Comment (henrygab) [reply src/usb/uf2/ghostfat.c]: ok, manually added those changes.   not sure why they got lost, I thought GIT would be smart enough to merge them properly.

**Comment by henrygab** [comment]
> Thank you very much for putting much thoughts into this PR regarding FAT file system. Sorry for late response, I was a bit busy with STM32 port for my tinyusb stack recently.

No problems on response time.  I'm thrilled to see the Adafruit implementation become the de-facto standard for how to implement UF2.

**Comment by hathach** [comment]
Superb !!! This looks perfect, thank you very much for putting much effort into this PR

**Comment by hathach** [comment]
@henrygab I am back working with bootloader today. The `ARRAY_SIZE2` is great, but this bootloader repo will be only compiled with C (unlikely to C++). Therefore the ARRAY_SIZE2 is overkill and may never get used. I am tempted to remove it, what do you think ?

**Comment by henrygab** [comment]
@hathach In the end, this is your repository, so it's your call.  However, since you ask, here is my _opinion_:

1. because the repo is intended to be compiled as C, remove the [warning](https://github.com/adafruit/Adafruit_nRF52_Bootloader/blob/2f314e1486f1a1e74ddd94e98df7decf3ea10594/src/usb/uf2/macros.h#L159) at line 159.
https://github.com/adafruit/Adafruit_nRF52_Bootloader/blob/2f314e1486f1a1e74ddd94e98df7decf3ea10594/src/usb/uf2/macros.h#L159

2. Keep the [first C++ solution](https://github.com/adafruit/Adafruit_nRF52_Bootloader/blob/2f314e1486f1a1e74ddd94e98df7decf3ea10594/src/usb/uf2/macros.h#L36-L121) (lines 26-121)

3. Remove the [other C++ solutions](https://github.com/adafruit/Adafruit_nRF52_Bootloader/blob/2f314e1486f1a1e74ddd94e98df7decf3ea10594/src/usb/uf2/macros.h#L122-L156) (lines 122-156)

This will simplify the code, while still getting benefits should the code become a baseline for any C++ work.  Also, while that first solution looks ugly, it works on compilers all the way back to Visual C++ 6(!).  This makes it elegant in my view.

Again, the repository is yours, and you have to maintain it, so of course the decision is yours.  I will not take offense, regardless of which way you decide to go.

**Comment by hathach** [comment]
@henrygab thank you for your suggestion. It is unlikely the bootloader is compiled with c++. All of source files are .c. I prefer to remove them all. It is useful for application such as Arduino Lib, not bootloader though. 🍺🍺

**Pull request merged**